contracts/hooks/examples/FullRange.sol

```diff
@@ -0,0 +1,360 @@
+    }
+
+    struct PoolInfo {
+        bool hasAccruedFees;
+        address liquidityToken;
+    }
+
+    struct AddLiquidityParams {
+        address token0;
+        address token1;
+        uint24 fee;
+        uint256 amount0Desired;
+        uint256 amount1Desired;
+        uint256 amount0Min;
+        uint256 amount1Min;
+        address to;
+        uint256 deadline;
+    }
+
+    struct RemoveLiquidityParams {
+        address token0;
+        address token1;
+        uint24 fee;
+        uint256 liquidity;
+        uint256 deadline;
+    }
+
+    mapping(PoolId => PoolInfo) public poolInfo;
+
+    constructor(IPoolManager _poolManager) BaseHook(_poolManager) {}
+
+    modifier ensure(uint256 deadline) {
+        require(deadline >= block.timestamp, "Expired");
+        _;
+    }
+
+    function getHooksCalls() public pure override returns (Hooks.Calls memory) {
+        return Hooks.Calls({
+            beforeInitialize: true,
+            afterInitialize: false,
+            beforeModifyPosition: true,
+            afterModifyPosition: false,
+            beforeSwap: true,
+            afterSwap: false,
+            beforeDonate: false,
+            afterDonate: false
+        });
+    }
+
+    function addLiquidity(AddLiquidityParams calldata params)
+        external
+        ensure(params.deadline)
+        returns (uint128 liquidity)
+    {
+        PoolKey memory key = PoolKey({
+            currency0: Currency.wrap(params.token0),
+            currency1: Currency.wrap(params.token1),
+            fee: params.fee,
+            tickSpacing: 60,
+            hooks: IHooks(address(this))
+        });
+
+        PoolId poolId = key.toId();
+
+        (uint160 sqrtPriceX96,,,,,) = poolManager.getSlot0(poolId);
+
+        if (sqrtPriceX96 == 0) revert PoolNotInitialized();
+
+        PoolInfo storage pool = poolInfo[poolId];
+
+        uint128 poolLiquidity = poolManager.getLiquidity(poolId);
+
+        liquidity = LiquidityAmounts.getLiquidityForAmounts(
+            sqrtPriceX96,
+            TickMath.getSqrtRatioAtTick(MIN_TICK),
+            TickMath.getSqrtRatioAtTick(MAX_TICK),
+            params.amount0Desired,
+            params.amount1Desired
+        );
+
+        if (poolLiquidity == 0 && liquidity <= MINIMUM_LIQUIDITY) {
+            revert LiquidityDoesntMeetMinimum();
+        }
+        BalanceDelta addedDelta = modifyPosition(
+            key,
+            IPoolManager.ModifyPositionParams({
+                tickLower: MIN_TICK,
+                tickUpper: MAX_TICK,
+                liquidityDelta: liquidity.toInt256()
+            })
+        );
+
+        if (poolLiquidity == 0) {
+            // permanently lock the first MINIMUM_LIQUIDITY tokens
+            liquidity -= MINIMUM_LIQUIDITY;
+            UniswapV4ERC20(pool.liquidityToken).mint(address(0), MINIMUM_LIQUIDITY);
+        }
+
+        UniswapV4ERC20(pool.liquidityToken).mint(params.to, liquidity);
+
+        require(
```

Comment:
nit: seems more appropriate on line 140?

Reply:
the issue is, if it's the first LP, then i'll mint them the liquidity without subtracting the amount of liquidity they are supposed to lock